pkg/streamingpromql/operators/functions/histograms.go

```diff
@@ -0,0 +1,346 @@
+			panic("We should never see more than one native histogram per group")
+		}
+		g.nativeHistograms = &hPoints
+	}
+	// hPoint's are returned to the pool after computeOutputSeriesForGroup is finished with them
+	g.remainingSeriesCount--
+}
+
+func (h *HistogramFunctionOverInstantVector) computeOutputSeriesForGroup(_ context.Context, g *bucketGroup) (types.InstantVectorSeriesData, error) {
+	floatPoints, err := types.FPointSlicePool.Get(h.timeRange.StepCount, h.memoryConsumptionTracker)
+	if err != nil {
+		return types.InstantVectorSeriesData{}, err
+	}
+
+	histogramIndex := 0
+	var currentHPoint *promql.HPoint
+
+	for pointIdx := range h.timeRange.StepCount {
+		ph := h.phValues.Samples[pointIdx].F
+		if math.IsNaN(ph) || ph < 0 || ph > 1 {
+			// Even when ph is invalid we still return a series as BucketQuantile will return +/-Inf.
+			// So don't skip/continue the series, just emit a warning.
+			h.Annotations.Add(annotations.NewInvalidQuantileWarning(ph, h.phArg.ExpressionPosition()))
+		}
+
+		// Get the HPoint if it exists at this step
+		if g.nativeHistograms != nil && histogramIndex < len(*g.nativeHistograms) {
+			nextHPoint := &(*g.nativeHistograms)[histogramIndex]
+			if h.timeRange.PointIndex(nextHPoint.T) == int64(pointIdx) {
+				currentHPoint = nextHPoint
+				histogramIndex++
+			}
+		}
+
+		if len(g.pointBuckets[pointIdx]) > 0 && currentHPoint != nil && h.timeRange.PointIndex(currentHPoint.T) == int64(pointIdx) {
+			// At this data point, we have classic histogram buckets and a native histogram with the same name and labels.
+			// No value is returned, so emit an annotation and continue.
+			h.Annotations.Add(annotations.NewMixedClassicNativeHistogramsWarning(
+				g.groupedMetricName, h.inner.ExpressionPosition(),
+			))
+			continue
+		}
+
+		if len(g.pointBuckets[pointIdx]) > 0 {
+			res, forcedMonotonicity, _ := bucketQuantile(ph, g.pointBuckets[pointIdx])
+			floatPoints = append(floatPoints, promql.FPoint{
+				T: h.timeRange.IndexTime(int64(pointIdx)),
+				F: res,
+			})
+			if forcedMonotonicity {
+				// Currently Prometheus does not correctly place the metric name onto this annotation.
+				// See: https://github.com/prometheus/prometheus/issues/15411
+				h.Annotations.Add(annotations.NewHistogramQuantileForcedMonotonicityInfo(
+					g.groupedMetricName, h.inner.ExpressionPosition(),
+				))
+			}
+			continue
+		}
+
+		if currentHPoint != nil && h.timeRange.PointIndex(currentHPoint.T) == int64(pointIdx) {
+			res := histogramQuantile(ph, currentHPoint.H)
+
+			floatPoints = append(floatPoints, promql.FPoint{
+				T: h.timeRange.IndexTime(int64(pointIdx)),
+				F: res,
+			})
+		}
+	}
+
+	// Return any retained native histogram to the pool
+	if g.nativeHistograms != nil {
+		types.HPointSlicePool.Put(*g.nativeHistograms, h.memoryConsumptionTracker)
+	}
+
+	return types.InstantVectorSeriesData{Floats: floatPoints}, nil
+}
+
+func (h *HistogramFunctionOverInstantVector) Close() {
+	h.inner.Close()
```

Comment:
We also need to call `Close` on the scalar operator.

Reply:
Good catch, thank you.